modules/mapper-extras/src/main/java/org/elasticsearch/index/query/ScoreScriptUtils.java

```diff
@@ -100,6 +147,122 @@ private static double intDotProduct(List<Number> v1, float[] v2){
 
 
     //**************FUNCTIONS FOR SPARSE VECTORS
+    /**
+     * Calculate l1 norm - Manhattan distance
+     * between a query's sparse vector and documents' sparse vectors
+     *
+     * L1NormSparse is implemented as a class to use
+     * painless script caching to prepare queryVector
+     * only once per script execution for all documents.
+     * A user will call `l1normSparse(params.queryVector, doc['my_vector'])`
+     */
+    public static final class L1NormSparse {
+        final double[] queryValues;
+        final int[] queryDims;
+
+        // prepare queryVector once per script execution
+        // queryVector represents a map of dimensions to values
+        public L1NormSparse(Map<String, Number> queryVector) {
+            //break vector into two arrays dims and values
+            int n = queryVector.size();
+            queryDims = new int[n];
+            queryValues = new double[n];
+            int i = 0;
+            for (Map.Entry<String, Number> dimValue : queryVector.entrySet()) {
+                try {
+                    queryDims[i] = Integer.parseInt(dimValue.getKey());
+                } catch (final NumberFormatException e) {
+                    throw new IllegalArgumentException("Failed to parse a query vector dimension, it must be an integer!", e);
+                }
+                queryValues[i] = dimValue.getValue().doubleValue();
+                i++;
+            }
+            // Sort dimensions in the ascending order and sort values in the same order as their corresponding dimensions
+            sortSparseDimsDoubleValues(queryDims, queryValues, n);
+        }
+
+        public double l1normSparse(VectorScriptDocValues.SparseVectorScriptDocValues dvs) {
+            BytesRef value = dvs.getEncodedValue();
+            if (value == null) return 0;
+            int[] docDims = VectorEncoderDecoder.decodeSparseVectorDims(value);
+            float[] docValues = VectorEncoderDecoder.decodeSparseVector(value);
+            int queryIndex = 0;
+            int docIndex = 0;
+            double l1norm = 0;
+            // find common dimensions among vectors v1 and v2 and calculate l1norm based on common dimensions
+            while (queryIndex < queryDims.length && docIndex < docDims.length) {
+                if (queryDims[queryIndex] == docDims[docIndex]) {
+                    l1norm += Math.abs(queryValues[queryIndex] - docValues[docIndex]);
+                    queryIndex++;
+                    docIndex++;
+                } else if (queryDims[queryIndex] > docDims[docIndex]) {
+                    docIndex++;
+                } else {
+                    queryIndex++;
+                }
```

Comment:
I have a general question about computing these norms for sparse vectors. The way this is implemented currently it looks like it is completely ignoring values that are not present in either of the two vectors. e.g. if a = [1:1, 2:1] and b = [3:1 4:1] then the result is 0. I would expect it to be 4 though. 
My assumtion (at least thats how I remember sparse vector behaviour from other libraries) is that each dimension that is not present is implicitely regarded as a 0, so it should be taken into account in the computation.

Reply:
@cbuescher  thanks, this is a great feedback. I will adjust the code accordingly